### PR TITLE
Static pages: implemented live preview

### LINKED ISSUE #3766: Add live preview when editing static documents
When editing static document, let's split the screen vertically if it's width is more than 1920px, and horizontally otherwise, and add a live rendering when contents of the HTML editor area change. In the preview mode, use the same wrapper divs to make sure we pick up all styles and render the page as close to the final output as possible.

Quick illustration: http://www.htmlinstant.com/ or http://htmledit.squarefree.com/

Mockup screenshot:
static_preview

### PR DESCRIPTION
**Pull request by julen**
This PR implements the live preview feature for static pages.

I'm putting this up for testing and further feedback. While at its current state this is fully functional, I'd still like to make a couple of changes before considering to merge it:
* Fix the styling of the content editing/previewing boxes and adjust their heights
* Make the `<ContentPreview />` component _dumb_ (i.e. do not hold any local state, just accept props)

Fixes #3766.

### CODE REVIEW

**Comment by dwaynebailey** [comment]
Works.  The live rendering is quite slow.  Might explain why github does it on another tab.

The other issue is that if you have content that is in any way long then the preview is not easily available as you need to scroll down.  IIRC the live preview was meant to be on the right.  But either on the right or below I can imagine space will always be an issue.

LGTM.

**Comment by dwaynebailey** [comment]
Oh I tested with HTML and Markdown, both are fine.  Its nice to have code highlighting in the content, helps prevent some errors.

**Comment by julen** [comment]
> The live rendering is quite slow

The server call to retrieve the rendered content is debounced at 500ms, I just adjusted it to 300ms.

> The other issue is that if you have content that is in any way long then the preview is not easily available as you need to scroll down.

This is one of the changes I want to make before merging as written above.

>  IIRC the live preview was meant to be on the right.

This depends on the viewport size. Currently the vertical split happens at widths larger than 1920px.

File: pootle/static/js/admin/general/components/ContentPreview.css [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (C) Pootle contributors.
+ *
+ * This file is a part of the Pootle project. It is distributed under the GPL3
+ * or later license. See the LICENSE file for a copy of the license and the
+ * AUTHORS file for copyright and authorship information.
+ */
```
Comment (unho) [inline pootle/static/js/admin/general/components/ContentPreview.css]: Do we put now the CSS files under the `js` directory?

**Comment by unho** [comment]
Did some quick testing and seems to work. I have found no issue that hasn't been already raised.

File: pootle/static/js/admin/general/components/ContentPreview.css (hunk 1 above)
Comment (julen) [inline pootle/static/js/admin/general/components/ContentPreview.css]: These are self-contained components, CSS and JS being a separation of technologies, not concerns. Concerns are being kept together.

**Comment by julen** [comment]
All the issues that I was mentioning in the first message have been ironed out and I think this is in good shape to be merged. The live preview component now also avoids the server round-trip when the markup is HTML, and in such scenario the preview is immediate.

**Comment by iafan** [comment]
@julen, is this live on stage?

**Comment by julen** [comment]
Yep.

**Comment by iafan** [comment]
Is it possible to display the "Allowed markup: HTML" and "Preview will be displayed here" placeholders as semi-transparent (gray) labels, aligned the same way (e.g. in the top left corner)?

**Comment by iafan** [comment]
Also: when calculating the height of the both panes in horizontal split mode, can we make them take into account the height of the page parts above and below? The idea is to avoid vertical scrolling so that the entire page conent fits the window size.

These lines can probably be merged into one and displayed to the right of the Save  buttons:
 ```
Use <a href="#/page-slug">Link text</a> code to provide links between static pages. Such links will be rewritten with full paths to static pages.

In Announcement pages use /xx/ within URLs for a dynamic language code.
```

**Comment by julen** [comment]
All the comments have now been addressed.

**Comment by iafan** [comment]
> The idea is to avoid vertical scrolling so that the entire page content fits the window size.

I believe this one is still not fixed (page height is still bigger than viewport height).

**Comment by julen** [comment]
It was requested for the horizontal split; are you checking the vertical split perhaps?

**Comment by iafan** [comment]
Yes, this split thing is confusing, sorry. When in *vertical* split (one pane on the left, and one on the right) we want the entire page to fit.

Having said that, in horizontal mode page also doesn't fit vertically (tested in Chrome).

**Comment by julen** [comment]
Can you share a screenshot and the screen resolution you have?

**Comment by iafan** [comment]
screenclip

screenclip 1

**Comment by iafan** [comment]
The first screenshot is from a full-screen mode. Display resolution is 2560x1440.

**Comment by julen** [comment]
There's probably some misunderstanding as from our PM you confirmed we want to have the entire *form* visible in the viewport, not the entire *page*.

**Comment by iafan** [comment]
I apologize. Entire page is what I had in mind. I want to avoid any page scrolling (unless the page is too small), because we have scrollable areas inside.

**Comment by julen** [comment]
Cool, all clear now.

**Comment by julen** [comment]
Implemented the requested changes (live on stage), and the resulting experience is much more pleasant and friendly — thanks for the great input, @iafan!

**Comment by iafan** [comment]
I tested this on stage, and though the edit/preview area height is smaller than before, the entire page doesn't fully fit:

screenclip

Tested in Chrome and in Firefox. In Firefox, the 'Save' button is visible, but the page footer (around 50 px) is still below the viewport and causes vertical scrolling.

**Comment by julen** [comment]
Fixed this and other tangential issues we discussed via IM.

**Comment by iafan** [comment]
GTM!

**Pull request merged**